Introduce RecordFunctionHolder to get rid of cpp_custom_type_hack:

bowangbj opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#60721 Introduce RecordFunctionHolder to get rid of cpp_custom_type_hack:**

1) Introduce RecordFunctionHolder and make it visible in Python.
2) modify record_function_enter return type from the hacky Tensor to RecordFunctionHolder
3) modify record_function_exit to take RecordFunctionHolder (instead of Tensor)

Summary:

Test Plan:

https://docs.google.com/document/d/1_ZcjIyIHKRbe1_dXYh7oTqTn0APEfnEEu1AySQwd3Qk/edit
Note still blocked by the jit.trace exception

Reviewers: rohan-varma, jerryzh168

Subscribers:

Tasks: T89059530

Tags:

tmp

Differential Revision: [D29386534](https://our.internmc.facebook.com/intern/diff/D29386534)

bowangbj committed:
This is first attempt to get rid of cpp_custom_type_hack.
To minimize the change, main changes:
1) Introduce RecordFunctionHolder and make it visible in Python.
2) modify record_function_enter return type from the hacky Tensor to RecordFunctionHolder
3) modify record_function_exit to take RecordFunctionHolder (instead of Tensor)

Alternative:
- In profiler, $ holder = torch.classes.profiler._RecordFunctionHolder()
- use holder.enter(name) and holder.exit()
- didn't choose b/c there are quite a few places which depend on current record_function_enter/exit ops

Follow up:
- merge record_function_enter_new and record_function_enter
- merge record_function_exit_new and record_function_exit
- modify C++ references of those two functions to use the clean interface.
- get rid of cpp_custom_type_hack.

Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

bowangbj committed:
Update on "This is first attempt to get rid of cpp_custom_type_hack."


To minimize the change, main changes:
1) Introduce RecordFunctionHolder and make it visible in Python.
2) modify record_function_enter return type from the hacky Tensor to RecordFunctionHolder
3) modify record_function_exit to take RecordFunctionHolder (instead of Tensor)

Alternative:
- In profiler, $ holder = torch.classes.profiler._RecordFunctionHolder()
- use holder.enter(name) and holder.exit()
- didn't choose b/c there are quite a few places which depend on current record_function_enter/exit ops

Follow up:
- merge record_function_enter_new and record_function_enter
- merge record_function_exit_new and record_function_exit
- modify C++ references of those two functions to use the clean interface.
- get rid of cpp_custom_type_hack.

Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

bowangbj commented:
@bowangbj has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29386534).

bowangbj committed:
local custom type hack: part I on "This is first attempt to get rid of cpp_custom_type_hack."


To minimize the change, main changes:
1) Introduce RecordFunctionHolder and make it visible in Python.
2) modify record_function_enter return type from the hacky Tensor to RecordFunctionHolder
3) modify record_function_exit to take RecordFunctionHolder (instead of Tensor)

Alternative:
- In profiler, $ holder = torch.classes.profiler._RecordFunctionHolder()
- use holder.enter(name) and holder.exit()
- didn't choose b/c there are quite a few places which depend on current record_function_enter/exit ops

Follow up:
- merge record_function_enter_new and record_function_enter
- merge record_function_exit_new and record_function_exit
- modify C++ references of those two functions to use the clean interface.
- get rid of cpp_custom_type_hack.

Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29386534](https://our.internmc.facebook.com/intern/diff/D29386534)

[ghstack-poisoned]

bowangbj committed:
local custom type hack: part I on "This is first attempt to get rid of cpp_custom_type_hack."


To minimize the change, main changes:
1) Introduce RecordFunctionHolder and make it visible in Python.
2) modify record_function_enter return type from the hacky Tensor to RecordFunctionHolder
3) modify record_function_exit to take RecordFunctionHolder (instead of Tensor)

Alternative:
- In profiler, $ holder = torch.classes.profiler._RecordFunctionHolder()
- use holder.enter(name) and holder.exit()
- didn't choose b/c there are quite a few places which depend on current record_function_enter/exit ops

Follow up:
- merge record_function_enter_new and record_function_enter
- merge record_function_exit_new and record_function_exit
- modify C++ references of those two functions to use the clean interface.
- get rid of cpp_custom_type_hack.

Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29386534](https://our.internmc.facebook.com/intern/diff/D29386534)

[ghstack-poisoned]

bowangbj committed:
local custom type hack: part I on "TESTTESTTEST  This is first attempt to get rid of cpp_custom_type_hack."


To minimize the change, main changes:
1) Introduce RecordFunctionHolder and make it visible in Python.
2) modify record_function_enter return type from the hacky Tensor to RecordFunctionHolder
3) modify record_function_exit to take RecordFunctionHolder (instead of Tensor)

Alternative:
- In profiler, $ holder = torch.classes.profiler._RecordFunctionHolder()
- use holder.enter(name) and holder.exit()
- didn't choose b/c there are quite a few places which depend on current record_function_enter/exit ops

Follow up:
- merge record_function_enter_new and record_function_enter
- merge record_function_exit_new and record_function_exit
- modify C++ references of those two functions to use the clean interface.
- get rid of cpp_custom_type_hack.

Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29386534](https://our.internmc.facebook.com/intern/diff/D29386534)

[ghstack-poisoned]

bowangbj committed:
local custom type hack: part I on "bowangbowangbowang This is first attempt to get rid of cpp_custom_type_hack."


Main changes:
1) Introduce RecordFunctionHolder and make it visible in Python.
2) modify record_function_enter return type from the hacky Tensor to RecordFunctionHolder
3) modify record_function_exit to take RecordFunctionHolder (instead of Tensor)

Alternative:
- In profiler, $ holder = torch.classes.profiler._RecordFunctionHolder()
- use holder.enter(name) and holder.exit()
- didn't choose b/c there are quite a few places which depend on current record_function_enter/exit ops

Follow up:
- merge record_function_enter_new and record_function_enter
- merge record_function_exit_new and record_function_exit
- modify C++ references of those two functions to use the clean interface.
- get rid of cpp_custom_type_hack.

Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29386534](https://our.internmc.facebook.com/intern/diff/D29386534)

[ghstack-poisoned]

bowangbj committed:
local custom type hack: part I on "Introduce RecordFunctionHolder to get rid of cpp_custom_type_hack."


Part I changes:

1) Introduce RecordFunctionHolder and make it visible in Python.
2) modify record_function_enter return type from the hacky Tensor to RecordFunctionHolder
3) modify record_function_exit to take RecordFunctionHolder (instead of Tensor)

Alternative: use RecordFunctionHolder obj directly inside python
pytorch % holder = torch.classes.profiler._RecordFunctionHolder()
pytorch % holder.enter(name) and holder.exit()
- didn't choose b/c there are quite a few places which depend on current record_function_enter/exit ops

Follow up:
- merge record_function_enter_new and record_function_enter
- merge record_function_exit_new and record_function_exit
- modify C++ references of those two functions to use the clean interface.
- get rid of cpp_custom_type_hack.

Summary:

Test Plan:
Verify with local test.py from torch/autograd/profiler.py#L582
pip uninstall torch --> python setup.py clean --> python setup.py develop --> python test.py --> the right metric shows up

Reviewers: rohan-varma, jerryzh168

Subscribers:

Tasks: T89059530

Tags:

Differential Revision: [D29386534](https://our.internmc.facebook.com/intern/diff/D29386534)

[ghstack-poisoned]

bowangbj commented:
Hey Rohan, Jerry

Sent out the very first draft for review. Please take a look. Thanks

bowangbj commented:
@bowangbj has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29386534).

bowangbj commented:
@bowangbj has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29386534).

bowangbj committed:
local custom type hack: part I on "Introduce RecordFunctionHolder to get rid of cpp_custom_type_hack."


Part I changes:

1) Introduce RecordFunctionHolder and make it visible in Python.
2) modify record_function_enter return type from the hacky Tensor to RecordFunctionHolder
3) modify record_function_exit to take RecordFunctionHolder (instead of Tensor)

Alternative: use RecordFunctionHolder obj directly inside python
pytorch % holder = torch.classes.profiler._RecordFunctionHolder()
pytorch % holder.enter(name) and holder.exit()
- didn't choose b/c there are quite a few places which depend on current record_function_enter/exit ops

Follow up:
- merge record_function_enter_new and record_function_enter
- merge record_function_exit_new and record_function_exit
- modify C++ references of those two functions to use the clean interface.
- get rid of cpp_custom_type_hack.

Summary:

Test Plan:
Verify with local test.py from torch/autograd/profiler.py#L582
pip uninstall torch --> python setup.py clean --> python setup.py develop --> python test.py --> the right metric shows up

Reviewers: rohan-varma, jerryzh168

Subscribers:

Tasks: T89059530

Tags:

Differential Revision: [D29386534](https://our.internmc.facebook.com/intern/diff/D29386534)

[ghstack-poisoned]

bowangbj commented:
@bowangbj has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29386534).

bowangbj committed:
Update on "Introduce RecordFunctionHolder to get rid of cpp_custom_type_hack."


Part I changes:

1) Introduce RecordFunctionHolder and make it visible in Python.
2) modify record_function_enter return type from the hacky Tensor to RecordFunctionHolder
3) modify record_function_exit to take RecordFunctionHolder (instead of Tensor)

Alternative: use RecordFunctionHolder obj directly inside python
pytorch % holder = torch.classes.profiler._RecordFunctionHolder()
pytorch % holder.enter(name) and holder.exit()
- didn't choose b/c there are quite a few places which depend on current record_function_enter/exit ops

Follow up:
- merge record_function_enter_new and record_function_enter
- merge record_function_exit_new and record_function_exit
- modify C++ references of those two functions to use the clean interface.
- get rid of cpp_custom_type_hack.

Summary:

Test Plan:
Verify with local test.py from torch/autograd/profiler.py#L582
pip uninstall torch --> python setup.py clean --> python setup.py develop --> python test.py --> the right metric shows up

Reviewers: rohan-varma, jerryzh168

Subscribers:

Tasks: T89059530

Tags:

Differential Revision: [D29386534](https://our.internmc.facebook.com/intern/diff/D29386534)

[ghstack-poisoned]

bowangbj commented:
@bowangbj has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29386534).

bowangbj committed:
Update on "Introduce RecordFunctionHolder to get rid of cpp_custom_type_hack."


Part I changes:

1) Introduce RecordFunctionHolder and make it visible in Python.
2) modify record_function_enter return type from the hacky Tensor to RecordFunctionHolder
3) modify record_function_exit to take RecordFunctionHolder (instead of Tensor)

Alternative: use RecordFunctionHolder obj directly inside python
pytorch % holder = torch.classes.profiler._RecordFunctionHolder()
pytorch % holder.enter(name) and holder.exit()
- didn't choose b/c there are quite a few places which depend on current record_function_enter/exit ops

Follow up:
- merge record_function_enter_new and record_function_enter
- merge record_function_exit_new and record_function_exit
- modify C++ references of those two functions to use the clean interface.
- get rid of cpp_custom_type_hack.

Summary:

Test Plan:
Verify with local test.py from torch/autograd/profiler.py#L582
pip uninstall torch --> python setup.py clean --> python setup.py develop --> python test.py --> the right metric shows up

Reviewers: rohan-varma, jerryzh168

Subscribers:

Tasks: T89059530

Tags:

Differential Revision: [D29386534](https://our.internmc.facebook.com/intern/diff/D29386534)

[ghstack-poisoned]

bowangbj commented:
@bowangbj has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29386534).

rohan-varma commented on torch/autograd/profiler.py:
```diff
@@ -606,9 +606,7 @@ def __init__(self, name: str):
-        # Stores underlying RecordFunction as a tensor. TODO: move to custom
-        # class (https://github.com/pytorch/pytorch/issues/35026).
-        self.handle: torch.Tensor = torch.zeros(1)
+        self.handle: torch.classes.profiler._RecordFunctionHolder()
```
Does this work? IIUC this just defines the type, we need to actually initialize it?

bowangbj replied:
Great catch.

rohan-varma commented on torch/csrc/autograd/record_function_ops.cpp:
```diff
@@ -14,6 +14,47 @@ namespace torch {
+  }
+  void enter(const std::string& name) {
+    if (record_function_ == nullptr) {
+      LOG(ERROR) << "record_function_ should never be NULL";
```
Let's use `TORCH_CHECK` so that we throw an informative error here.

bowangbj replied:
Done

rohan-varma commented on torch/csrc/autograd/record_function_ops.cpp:
```diff
@@ -14,6 +14,47 @@ namespace torch {
+
+  void exit() {
+    if (record_function_ == nullptr) {
+      LOG(ERROR) << "record_function_ should never be NULL!";
```
same comment as above

bowangbj replied:
Done

rohan-varma reviewed (commented):
This is a great start, thanks for working on it! As a next step, can we add some tests to ensure everything works appropriately, and also replace the callsites that use `at::cpp_custom_type_hack` with the new APIs?

bowangbj committed:
resolve rohan comment, deprecate old record func on "Introduce RecordFunctionHolder to get rid of cpp_custom_type_hack."


Part I changes:

1) Introduce RecordFunctionHolder and make it visible in Python.
2) modify record_function_enter return type from the hacky Tensor to RecordFunctionHolder
3) modify record_function_exit to take RecordFunctionHolder (instead of Tensor)

Summary:

Test Plan:

https://docs.google.com/document/d/1_ZcjIyIHKRbe1_dXYh7oTqTn0APEfnEEu1AySQwd3Qk/edit

Two pending issues:1) unknown type from jit and 2) test_profiler_propagation failing

Reviewers: rohan-varma, jerryzh168

Subscribers:

Tasks: T89059530

Tags:

Differential Revision: [D29386534](https://our.internmc.facebook.com/intern/diff/D29386534)

[ghstack-poisoned]

bowangbj commented:
@bowangbj has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29386534).

bowangbj commented:
> This is a great start, thanks for working on it! As a next step, can we add some tests to ensure everything works appropriately, and also replace the callsites that use `at::cpp_custom_type_hack` with the new APIs?

Thanks Rohan for all your great comments. Resolved all of them. 
For the tests, since it's an internal change of the record function, I was mainly reusing existing test cases. I do found two failures because of the change: 

1. test_profiler_propagation fails because of an assertion error raised by /aten/src/ATen/core/builtin_function.h
2. test_tensorpipe_agent fails because Unknown type name 'torch.classes.profiler._RecordFunctionHolder thrown by torch._C._jit_script_compile

Please see 4 and 5 from this doc: https://docs.google.com/document/d/1_ZcjIyIHKRbe1_dXYh7oTqTn0APEfnEEu1AySQwd3Qk/edit#heading=h.suc7fabk5gdn

Can you take a quick look and give me some hint how to resolve the Unknown type issue and the assertion error at test_profiler_propagation 

Thanks a lot

bowangbj committed:
fix untype error and jit exception on "Introduce RecordFunctionHolder to get rid of cpp_custom_type_hack:"


1) Introduce RecordFunctionHolder and make it visible in Python.
2) modify record_function_enter return type from the hacky Tensor to RecordFunctionHolder
3) modify record_function_exit to take RecordFunctionHolder (instead of Tensor)

Summary:

Test Plan:

https://docs.google.com/document/d/1_ZcjIyIHKRbe1_dXYh7oTqTn0APEfnEEu1AySQwd3Qk/edit

Reviewers: rohan-varma, jerryzh168

Subscribers:

Tasks: T89059530

Tags:

Differential Revision: [D29386534](https://our.internmc.facebook.com/intern/diff/D29386534)

[ghstack-poisoned]

bowangbj commented:
@bowangbj has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29386534).

bowangbj commented:
> > This is a great start, thanks for working on it! As a next step, can we add some tests to ensure everything works appropriately, and also replace the callsites that use `at::cpp_custom_type_hack` with the new APIs?
> 
> Thanks Rohan for all your great comments. Resolved all of them.
> For the tests, since it's an internal change of the record function, I was mainly reusing existing test cases. I do found two failures because of the change:
> 
> 1. test_profiler_propagation fails because of an assertion error raised by /aten/src/ATen/core/builtin_function.h
> 2. test_tensorpipe_agent fails because Unknown type name 'torch.classes.profiler._RecordFunctionHolder thrown by torch._C._jit_script_compile
> 
> Please see 4 and 5 from this doc: https://docs.google.com/document/d/1_ZcjIyIHKRbe1_dXYh7oTqTn0APEfnEEu1AySQwd3Qk/edit#heading=h.suc7fabk5gdn
> 
> Can you take a quick look and give me some hint how to resolve the Unknown type issue and the assertion error at test_profiler_propagation
> 
> Thanks a lot

It turns out there is some magic how JIT interprets python code. I made a small tweak and resolved the JIT exception: 

self.handle : __torch__.torch.classes.profiler._RecordFunctionHolder = torch.ops.profiler._record_function_enter('placeholder')

note: 
1) self.handle : torch.classes.profiler._RecordFunctionHolder = torch.ops.profiler._record_function_enter('placeholder') will introduce Unknown Type error, I guess the fully qualified type is needed to let C++ know the type of the variable
2) torch.classes.profiler._RecordFunctionHolder() -->  torch.ops.profiler._record_function_enter('placeholder') to avoid JIT exception in test_profiler_propagation 

test_profiler_propagation still fails, @rohan-varma , could it be possible it was caused by the fact I am using intrusive_ptr ? 

Thanks

bowangbj committed:
clean up CMakeList.txt on "Introduce RecordFunctionHolder to get rid of cpp_custom_type_hack:"


1) Introduce RecordFunctionHolder and make it visible in Python.
2) modify record_function_enter return type from the hacky Tensor to RecordFunctionHolder
3) modify record_function_exit to take RecordFunctionHolder (instead of Tensor)

Summary:

Test Plan:

https://docs.google.com/document/d/1_ZcjIyIHKRbe1_dXYh7oTqTn0APEfnEEu1AySQwd3Qk/edit

Reviewers: rohan-varma, jerryzh168

Subscribers:

Tasks: T89059530

Tags:

Differential Revision: [D29386534](https://our.internmc.facebook.com/intern/diff/D29386534)

[ghstack-poisoned]

bowangbj commented:
@bowangbj has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29386534).

rohan-varma commented on torch/autograd/profiler.py:
```diff
@@ -606,9 +605,7 @@ def __init__(self, name: str):
-        # Stores underlying RecordFunction as a tensor. TODO: move to custom
-        # class (https://github.com/pytorch/pytorch/issues/35026).
-        self.handle: torch.Tensor = torch.zeros(1)
+        self.handle : __torch__.torch.classes.profiler._RecordFunctionHolder = torch.ops.profiler._record_function_enter('placeholder')
```
We should only call `_record_function_enter` in the `__enter__` part of the context manager, not during init time right? Here we should probably just initialize `self.handle: __torch__.classes.profiler.RecordFunctionHolder = RecordFunctionHolder()`? This may also fix `test_profiler_propagation`.

bowangbj replied:
Unfortunately seems if we use CustomClassObject.member_function, torch.jit.trace will throw an exception (the same as before), pls see https://docs.google.com/document/d/1_ZcjIyIHKRbe1_dXYh7oTqTn0APEfnEEu1AySQwd3Qk/edit#heading=h.si9n17vbfpv1 for context

bowangbj replied:
Anyway changed back to member function to unhide the potential JIT issue.

rohan-varma commented on torch/csrc/autograd/record_function_ops.cpp:
```diff
@@ -14,42 +14,44 @@ namespace torch {
-  auto& rec = at::cpp_custom_type_hack::cast<at::RecordFunction>(handle);
-  return rec;
+void RecordFunctionHolder::exit() {
+  TORCH_CHECK(record_function_ != nullptr,
```
We usually use TORCH_INTERNAL_ASSERT for bugs that are due to developer coding logic, and TORCH_CHECK for issues that are due to incorrect use of pytorch by the user. Since it doesn't appear that the user can get the application in this state, we should probably switch it to the former.

Also if it incurs a lot of overhead, we can switch to a debug check and it will be stripped out in non-debug builds.

bowangbj replied:
Got it

rohan-varma reviewed (commented):
Awesome! Just a few smaller comments and we still need to debug the `test_profiler_propagation` issue.

bowangbj committed:
resolve rohan comment on "Introduce RecordFunctionHolder to get rid of cpp_custom_type_hack:"


1) Introduce RecordFunctionHolder and make it visible in Python.
2) modify record_function_enter return type from the hacky Tensor to RecordFunctionHolder
3) modify record_function_exit to take RecordFunctionHolder (instead of Tensor)

Summary:

Test Plan:

https://docs.google.com/document/d/1_ZcjIyIHKRbe1_dXYh7oTqTn0APEfnEEu1AySQwd3Qk/edit
Note still blocked by the jit.trace exception

Reviewers: rohan-varma, jerryzh168

Subscribers:

Tasks: T89059530

Tags:

tmp

Differential Revision: [D29386534](https://our.internmc.facebook.com/intern/diff/D29386534)

[ghstack-poisoned]

bowangbj commented:
@bowangbj has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29386534).

bowangbj commented:
> Awesome! Just a few smaller comments and we still need to debug the `test_profiler_propagation` issue.

Thanks Rohan for your great review, updated per all your comments. Did two quick experiments: 

1) Did some hack to wrap existing changes inside a pure placeholder RecordFunctionHolder, verified that it seems the use of intrusive_ptr might be the culprit of breaking test_profiler_propagation. Do you happen to know some alternative of exposing CustomClass object in custom operator w/o using intrusive ptr? 

2) Attempted to use CustomClassObjection.method directly to verify whether we can resolve test_profiler_propagation w/o using custom operator, but got the torch.jit.trace exception as documented here: https://docs.google.com/document/d/1_ZcjIyIHKRbe1_dXYh7oTqTn0APEfnEEu1AySQwd3Qk/edit#heading=h.si9n17vbfpv1

PTAL and let me know if any trick to resolve the exception.
